feedbot support

ironm00n opened the pull request:
resolves: #2

ironm00n committed:
wip: feedbot scaffolding

dbp commented:
This is great, thanks :)

dbp committed:
add rate limit, missing params, default a lot more, placeholder prompt

dbp committed:
fix typo, add some tests

ironm00n committed:
test: add simple shim to run integration tests

dbp committed:
fix extraction of feedbot; strip info from trace afterwards; fix serialization

dbp committed:
real prompt for feedback, better shim

ironm00n commented:
> the fact that info is sometimes `Nothing`, sometimes `Some(T)`, is somewhat tricky. What I'm doing will not type check, I'm sure, but it works fine :)

What you did is correct, the type is morally $\exists T. T$ which we are just reflecting on.

dbp commented:
Does this code look good to merge? The corresponding PR on the pawtograder side has been merged, so it'd be good to get this merged if there aren't issues that should be addressed.

dbp committed:
add a bunch more (optional) arguments for feedbot grader

ironm00n committed:
test: also print entire json

ironm00n committed:
tweak feedbot prompt to include codeblock

ironm00n committed:
test: include updated feedbot params

ironm00n committed:
use sliced program in feedbot

`slice-from-function` currently is identity function but shouldn't need to change once properly implemented.
